### PR TITLE
Add OSF Harvester and Normalizer

### PR DESCRIPTION
**Pull request by erinspace**
Havester and normalizer for OSF!

### CODE REVIEW

File: providers/io/osf/harvester.py [hunk 1]
```diff
@@ -1,5 +1,70 @@
...
 
+logger = logging.getLogger(__name__)
+
 
 class OSFHarvester(Harvester):
-    pass
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.url = 'https://api.osf.io/v2/nodes/?page[size]=100&filter[public]=true&embed=affiliated_institutions'
```
Comment (chrisseto) [inline providers/io/osf/harvester.py]: Even if this is always the base URL use furl to construct the parameters

File: providers/io/osf/harvester.py [hunk 2]
```diff
@@ -1,5 +1,70 @@
...
+    def do_harvest(self, start_date: arrow.Arrow, end_date: arrow.Arrow) -> Iterator[Tuple[str, Union[str, dict, bytes]]]:
+
+        url = furl(self.url)
+
+        url.args['filter[date_modified][gt]'] = start_date.date().isoformat()
+        url.args['filter[date_modified][lt]'] = end_date.date().isoformat()
+
+        return self.fetch_records(url)
+
+    def fetch_records(self, url: furl) -> list:
```
Comment (chrisseto) [inline providers/io/osf/harvester.py]: The type annotation here should be the same as `do_harvest`

File: share/normalize/parsers.py [hunk 3]
```diff
@@ -47,7 +47,7 @@ def is_list(self):
 
     @property
     def parser(self):
-        # Peak into the module where all the parsers are being imported from
+        # Peek into the module where all the parsers are being imported from
```
Comment (chrisseto) [inline share/normalize/parsers.py]: 🍆

File: providers/io/osf/harvester.py [hunk 4]
```diff
@@ -1,5 +1,73 @@
+import arrow
+import logging
+
+from furl import furl
```
Comment (chrisseto) [inline providers/io/osf/harvester.py]: `arrow` is a 3rd party
`typing` is in the standard lib

File: providers/io/osf/normalizer.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        registrations = ctx.relationships.registrations.links.related.href
+        logs = ctx.relationships.logs.links.related.href
+        node_links = ctx.relationships.node_links.links.related.href
+        wikis = ctx.relationships.wikis.links.related.href
+        children = ctx.relationships.children.links.related.href
+        fork = ctx.attributes.fork
+        date_modified = ctx.attributes.date_modified
+        collection = ctx.attributes.collection
+        registration = ctx.attributes.registration
+        node_license = ctx.attributes.node_license
```
Comment (chrisseto) [inline providers/io/osf/normalizer.py]: This should be the `rights` field on CreativeWork

File: providers/io/osf/normalizer.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+        users = ctx.relationships.users.links.related.href
+        registrations = ctx.relationships.registrations.links.related.href
+        description = ctx.attributes.description
+
+
+class ThroughInstitutions(Parser):
+    institution = Delegate(ctx, Institution)
+
+
+class CreativeWork(Parser):
```
Comment (chrisseto) [inline providers/io/osf/normalizer.py]: Does the api provide a link back to the project?
Do we ever return DOIs or ARK ids?

File: providers/io/osf/harvester.py [hunk 7]
```diff
@@ -1,5 +1,73 @@
...
+                    contributor_data = contributor_data + contributor_records.json()['data']
+                    if not next_contributor_page:
+                        break
+                    contributor_records, next_contributor_page = self.fetch_page(next_contributor_page)
+                logger.info('Had {} contributors to harvest, harvested {}'.format(total_contributors, len(contributor_data)))
+                record['contributors'] = contributor_data
+
+                # gather the the rest of the record
+                total_harvested += 1
+                yield(record['id'], record)
```
Comment (chrisseto) [inline providers/io/osf/harvester.py]: space between `yield` and `(`

File: providers/io/osf/normalizer.py (hunk 6 above)
Comment (erinspace) [inline providers/io/osf/normalizer.py]: @chrisseto  Yes on the link, added!

NO on the DOIs for Projects, YES on them for Registrations. Creating a separate provider for OSF Registrations....

**Comment by erinspace** [comment]
@chrisseto 🌴

**Pull request merged**